parsec/mca/device/device_gpu.c

```diff
@@ -1949,6 +1957,20 @@ parsec_device_progress_stream( parsec_device_gpu_module_t* gpu_device,
 
   schedule_task:
     rc = progress_fct( gpu_device, task, stream );
+    if(0 == rc) {
+#if defined(PARSEC_PROF_TRACE)
+        if( stream->prof_event_track_enable ) {
+            if( task->prof_key_end != -1 ) {
+                PARSEC_PROFILING_TRACE(stream->profiling, task->prof_key_end, task->prof_event_id, task->prof_tp_id, NULL);
+            }
+        }
+#endif
+        /* If progress_fct added nothing on that stream, we skip scheduling a record on the GPU stream */
+        if( task->complete_stage )
+            rc = task->complete_stage(gpu_device, &task, stream);
+        *out_task = task;
+        return rc;
+    }
```

Comment:
How about we rely on the already existing code for returning a completed task from the stream that is a little above in this function. Instead of replicating the code here, and creating two identical return points (both returning the task can be advanced to the next stage), we merge these two into a single. For this we can replace the code above with
```c
task->last_status = PARSEC_HOOK_RETURN_DONE;
goto withdraw_the_task;
```
and then add the `withdraw_the_task` label at line 1910 in my version of the file, basically right after `stream->end = (stream->end + 1) % stream->max_events;`.

Reply:
But the code below 1910 does modify `stream->tasks`, which we should not touch here. Plus, it won't help readability.